### PR TITLE
OSDOC-1984 Azure CSI

### PR DESCRIPTION
**Pull request by lpettyjo**
Azure CSI feature for 4.8 OCP
OSDOCS: https://issues.redhat.com/browse/OSDOCS-1984
JIRA epic: https://issues.redhat.com/browse/STOR-364
Preview: https://deploy-preview-31491--osdocs.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-azure.html

@bobfuru PTAL.
@qinpingli @huffmanca PTAL.

### CODE REVIEW

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:17]: I suggest commenting out your question so it doesn't display in preview output or inadvertently later in published output. :)

File: _topic_map.yml
Comment (bobfuru) [inline _topic_map.yml:1068]: This should be renamed "Azure **Disk** CSI Driver Operator" to distinguish from [Azure **File** CSI](https://issues.redhat.com/browse/STOR-554), which this feature does not cover.

File: modules/persistent-storage-csi-drivers-supported.adoc
Comment (bobfuru) [inline modules/persistent-storage-csi-drivers-supported.adoc:20]: To match the name of the driver, either:
s/Microsoft Azure/Azure Disk/ or s/Microsoft Azure/Microsoft Azure Disk/
(If the latter, then move below Google for alpha ordering.)

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:2]: s/Azure CSI/Azure Disk CSI/

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:1]: s/[id="persistent-storage-csi-azure"]/[id="persistent-storage-csi-azure-disk"]

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:4]: s/persistent-storage-csi-azure/persistent-storage-csi-azure-disk

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:10]: In this first sentence only, I suggest writing out "Microsoft Azure Disk Storage" to match their latest terminology per: https://azure.microsoft.com/en-us/services/storage/disks/
For the rest of the doc, I think just "Azure Disk" more closely matches our CSI driver terminology.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:12]: s/Azure Driver Operator/Azure Disk CSI Driver Operator

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:17]: - s/Azure storage assets/Azure Disk storage assets/
- s/Azure CSI Driver Operator and the Azure CSI driver/Azure Disk CSI Driver Operator and the Azure Disk CSI driver

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:19]: extra line break

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:21]: s/Azure CSI/Azure Disk CSI/

Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:21]: We recently updated how to format object references in our guidelines: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#object-references

Because of this, you can change the following:
- s/StorageClass by default/storage class by default/
- s/create the Azure StorageClass as described/create the Azure Disk storage class as described

I did notice that this guidance wasn't applied in [AWS EBS](https://docs.openshift.com/container-platform/4.7/storage/container_storage_interface/persistent-storage-csi-ebs.html) docs, so that should probably be changed at some point as well.

Also, since "PVCs" wasn't defined before:

- s/to create PVCs./to create persistent volume claims (PVCs).

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:23]: - s/Azure CSI driver/Azure Disk CSI driver/
- s/Azure PVs/Azure Disk PVs/ (<--I think?)

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:29]: Just to confirm, [Azure Disk migration](https://issues.redhat.com/browse/STOR-576) is not targeted for 4.8, correct?

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:32]: s/Azure persistent volumes/Azure Disk persistent volumes/
s/Persistent storage using Azure Block Store/Persistent storage using Azure Disk/
^^ I just noticed the h1 in this assembly (`persistent-storage-azure.adoc`) should also probably be changed to "Azure Disk" to distinguish from the separate "Azure File" doc :)

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:35]: s/Persistent storage using Azure Block Store/Persistent storage using Azure Disk/

**Review by bobfuru — COMMENTED** [review]
Most of this is in relation to clarifying that this is for Azure Disk CSI (as opposed to Azure File CSI or Azure Disk in-tree). Let me know if you have questions about any of this.

Also, please add the preview link and the corresponding Jira link in the first comment/description for easier reference.

I didn't mark as "peer review done" because this still needs clarification about the installation instructions.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (lpettyjo) [reply storage/container_storage_interface/persistent-storage-csi-azure.adoc:29]: Correct. Only Cinder and AWS EBS are in scope for migration in 4.8.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (huffmanca) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:17]: I'd be careful here, as the CSI Driver and Operator aren't installed by default. The wording implies that it is. The administrator must enable the `TechPreviewNoUpgrade` feature gate to have access to this driver and operator.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (huffmanca) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:30]: Should we include those steps here as well, using the `managed-csi` StorageClass that's installed?

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (huffmanca) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:19]: I would reword the second sentence here:

> You can create additional storage classes, as described in $LINK, to accommodate various storage profiles.

File: modules/persistent-storage-csi-drivers-supported.adoc
Comment (huffmanca) [inline modules/persistent-storage-csi-drivers-supported.adoc:21]: The Azure Disk CSI Driver supports snapshots, cloning, and resizing.

**Review by huffmanca — CHANGES_REQUESTED** [review]

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (qinpingli) [reply storage/container_storage_interface/persistent-storage-csi-azure.adoc:17]: Agree with @huffmanca , I'd like to add how to install the csi driver and operator here. And add a note about if we used the `TechPreviewNoUpgrade` featureset to install csi driver and operator, it's undone and will block the upgrade.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (qinpingli) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:27]: s/provision Azure storage/provision Azure disk storage/ ?

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (qinpingli) [reply storage/container_storage_interface/persistent-storage-csi-azure.adoc:19]: The internal provisioner is used in the $LINK, how about adding some description here, for the CSI driver, we should use the `disk.csi.azure.com` provisioner?

In case the CSI driver has different parameters from in-tree plugin, maybe we can add more reference doc: https://github.com/openshift/azure-disk-csi-driver/blob/master/docs/driver-parameters.md?

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (huffmanca) [inline modules/persistent-storage-csi-tp-enable.adoc:15]: Should we be specific about the feature gate that needs to be enabled?

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (huffmanca) [inline modules/persistent-storage-csi-tp-enable.adoc:19]: Same comment here - should we be explicit that it's the `TechPreviewNoUpgrade` feature gate? I see multiple listed in the referenced assembly.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (huffmanca) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:17]: It looks like we're still saying `by default` here.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (huffmanca) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:19]: Same comment about `by default` here. Perhaps we can say, "Once this feature is enabled, a storage class is created by default...` ?

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (lpettyjo) [reply modules/persistent-storage-csi-tp-enable.adoc:15]: Yup. Done!

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (lpettyjo) [reply modules/persistent-storage-csi-tp-enable.adoc:19]: Yes!

File: modules/persistent-storage-csi-drivers-supported.adoc
Comment (duanwei33) [inline modules/persistent-storage-csi-drivers-supported.adoc:21]: Do we need to remove the "Microsoft" here? @jsafrane 
Also, Azure don't support the clone feature yet.

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (duanwei33) [inline modules/persistent-storage-csi-tp-enable.adoc:59]: Did you install the upstream CSI driver? we should have the pod with:
azure-disk-csi-driver-controller-5949bf45fd-pm4qb 11/11 Running 0 39m
azure-disk-csi-driver-node-2tcxr 3/3 Running 0 53m
azure-disk-csi-driver-node-2xjzm 3/3 Running 0 53m
azure-disk-csi-driver-node-6wrgk 3/3 Running 0 53m
azure-disk-csi-driver-node-frvx2 3/3 Running 0 53m
azure-disk-csi-driver-node-lf5kb 3/3 Running 0 53m
azure-disk-csi-driver-node-mqdhh 3/3 Running 0 53m
azure-disk-csi-driver-operator-7d966fc6c5-x74x5 1/1 Running 0 44m

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (duanwei33) [inline modules/persistent-storage-csi-tp-enable.adoc:98]: The Azure Disk in-tree storage class name should be "managed-premium", "thin" is for vsphere.
managed-premium (default) kubernetes.io/azure-disk Delete WaitForFirstConsumer true 76m

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (duanwei33) [inline modules/persistent-storage-csi-tp-enable.adoc:99]: Our Azure CSI driver operator installed the storage class with "disk.csi.azure.com" as PROVISIONER, not "csi.azure.microsoft.com"
managed-csi disk.csi.azure.com Delete WaitForFirstConsumer true 51m

File: modules/persistent-storage-csi-drivers-supported.adoc
Comment (bobfuru) [reply modules/persistent-storage-csi-drivers-supported.adoc:21]: Including "Microsoft" here is consistent with the other entries in this table. But I am still seeing cloning as supported here. Is that correct?

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:12]: s/Azure Disk Driver Operator/Azure Disk CSI Driver Operator

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:17]: Technically, the sentence seems correct but the commas/clauses trip me up a bit. Perhaps a bit of a rearrangement might help? Something like:
> To create CSI-provisioned PVs that mount to Azure Disk storage assets with this feature enabled, {product-title} installs the Azure Disk CSI Driver Operator and the Azure Disk CSI driver by default in the `openshift-cluster-csi-drivers` namespace.

Or could you just leave out the "this feature is enabled" part altogether?

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:19]: Same question, could you leave out "after being enabled"? This is just describing the feature anyway. 🤷

Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:19]: ISG suggests avoiding parentheses for clarity. (Not my favorite rule!)
So maybe:
s/a storage class (`managed-csi`)/a storage class called `managed-csi`?

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [inline storage/container_storage_interface/persistent-storage-csi-azure.adoc:31]:
```suggestion
In future {product-title} versions, volumes provisioned using existing in-tree plug-ins are planned for migration to their equivalent CSI driver. CSI automatic migration should be seamless. Migration will not change how you use all existing API objects, such as persistent volumes, persistent volume claims, and storage classes. For more information about migration, see xref:../../storage/container_storage_interface/persistent-storage-csi-migration.adoc#persistent-storage-csi-migration[CSI automatic migration].

After full migration, in-tree plug-ins will be removed in future versions of {product-title}.
```

^ A few suggestions for readability and ISG/docs guidelines (avoid commas and use lowercase for general API refs). If you agree with the changes, you'll need to apply manually...avoid applying the suggested change directly in GH.

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (bobfuru) [inline modules/persistent-storage-csi-tp-enable.adoc:14]: s/driver operator/Driver Operator

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (bobfuru) [inline modules/persistent-storage-csi-tp-enable.adoc:15]: s/operator/Operator

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (bobfuru) [inline modules/persistent-storage-csi-tp-enable.adoc:19]: You'll need to remove the xref as they can't be included in modules yet. Suggest something like, "See "Enabling features using feature gates" and then adding an xref in the assembly file under "Additional resources".

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (bobfuru) [inline modules/persistent-storage-csi-tp-enable.adoc:23]: s/on using/on by using

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (bobfuru) [inline modules/persistent-storage-csi-tp-enable.adoc:42]: You have "verify" in the sentence twice. Suggest rewording to something like, "Check the status..."

File: modules/persistent-storage-csi-tp-enable.adoc
Comment (bobfuru) [reply modules/persistent-storage-csi-tp-enable.adoc:59]: Should Wei Duan's values be used in place of lines 67-73? Those don't have `microsoft-` prepended to the filenames as yours do.

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (bobfuru) [reply storage/container_storage_interface/persistent-storage-csi-azure.adoc:4]: Looks like this was missed in the `context` field.

**Review by bobfuru — COMMENTED** [review]
Left some comments for style guide consistency and other wording suggestions for your consideration, along with an xref in a module to address. Additionally, the QE comments haven't been resolved yet so this should get another review from QE before it's ready for merge.

File: modules/persistent-storage-csi-drivers-supported.adoc
Comment (jsafrane) [reply modules/persistent-storage-csi-drivers-supported.adoc:21]: @duanwei33, upstream documents volume cloning as supported: https://github.com/kubernetes-sigs/azuredisk-csi-driver/tree/master/deploy/example/cloning
Can you please give it a quick test?

Comment (lpettyjo) [reply modules/persistent-storage-csi-drivers-supported.adoc:21]: @duanwei33 can you confirm that Azure cloning has been tested and doesn't work?

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (lpettyjo) [reply storage/container_storage_interface/persistent-storage-csi-azure.adoc:17]: Nope! You cannot leave out the "feature enabled" part. Engr/QE were very definite that I had to stress that the feature has to be enabled or you don't get anything. I used your first option. Thanks!

File: storage/container_storage_interface/persistent-storage-csi-azure.adoc
Comment (lpettyjo) [reply storage/container_storage_interface/persistent-storage-csi-azure.adoc:19]: Nope, sadly. As I mentioned above QE/Engr really want to stress the enabling part because this is new for storage. If you don't enable TP, you don't get the feature. They are very worried users will trip over this.

File: modules/persistent-storage-csi-drivers-supported.adoc
Comment (lpettyjo) [reply modules/persistent-storage-csi-drivers-supported.adoc:21]: Wei responded that we can consider Azure CSI as supporting cloning.

<br class="Apple-interchange-newline">
Wei Duan | &nbsp; | 2:34 AM (5 hours ago)
-- | -- | --
Hi Jan, As we have some discussion in slack, I think we can go with "azure clone supported" although there is AZ issue to be fixed | Hi Jan, As we have some discussion in slack, I think we can go with "azure clone supported" although there is AZ issue to be fixed

**Comment by bobfuru** [comment]
LGTM!

**Comment by bobfuru** [comment]
/cherrypick-enterprise-4.8

**Pull request merged**
